rfcs/002 - Side-effects in Marsha.md

```diff
@@ -0,0 +1,126 @@
+* cnn('world') = ["Police raids follow shocking video of sexual assault in India’s Manipur state amid ethnic violence", 'Ukrainian air defenses in Odesa outgunned as Russia targets global grain supply', 'Anger boils over as Kenya’s cost of living protests shake the nation']
+```
+
+Presumably the text in the current example explaining to the LLM that the headline class name has changed and what it now is would not be necessary because the mock would be copied from the reality of the website itself.
+
+This particular syntax might be confusing to non-developers, though, since the difference between a function input and output and a function side-effect source and destination could be very unclear, especially if "side-effect" is not in the nomenclature, but it would be brief as a syntax.
+
+#### Optional side-effect "arguments"
+
+Treating the side-effect inputs and outputs as non-side-effect inputs and outputs that are automatically set (or ignored for the return type) might make the concept less confusing. Wrapping the side-effects in both in brackets at the end of the lists could help here:
+
+```md
+# func cnn(string of section to take headlines from, [https://cnn.com]): list of headlines
+
+This function scrapes the cnn.com website for headlines. The section it takes the headlines from is passed to it, with 'home' referring to the homepage at cnn.com and should be special cased, while 'us' refers to cnn.com/us, 'politics' refers to cnn.com/politics, and so on for every top-level category CNN has.
+
+* cnn('home') = ["Florida's new standards for teaching Black history spark outrage", "His books sold over 300 million copies and were translated into 63 languages. Now, a museum is acknowledging his racism", "Player quits match in tears as tennis world slams opponent’s ‘absolutely disgusting’ actions"]
+* cnn('us') = ['18-year-old Miami woman arrested after allegedly trying to hire a hitman to go after her 3-year-old son', 'Investigation into Gilgo Beach serial killings suspect expands to Nevada and South Carolina', 'Rescue crews continue search for 2 children swept away by Pennsylvania floodwater that killed their mother']
+* cnn('world') = ["Police raids follow shocking video of sexual assault in India’s Manipur state amid ethnic violence", 'Ukrainian air defenses in Odesa outgunned as Russia targets global grain supply', 'Anger boils over as Kenya’s cost of living protests shake the nation']
+```
+
+Here it's clear that the `[https://cnn.com]` is *different*, though exactly how is not clear on a first reading. The fact that this input argument doesn't show up in the input arguments list of the generated function but it does show up in the function body might help clear it up a bit for the user just from inspection, but it could also confuse people that it's a "default" value of some sort that you can replace with some other value when that is not the case.
```

Comment:
Also `[]` are sometimes used as optional values and that could be confusing. This alternative would not take into account the "output" side effect either, right?

Reply:
This is one of the suggestions I preferred, though, even though I agree this particular syntax is confusing. It doesn't help with discoverability, but it does make it obvious that it is an explicitly added feature in Marsha as a project with implications on how the compilation will go, rather than just relying on the LLM to "do the right thing" as some of the other examples imply.